ci/publish.rs

```diff
@@ -0,0 +1,242 @@
+fn find_crates(dir: &Path, dst: &mut Vec<Crate>) {
+    if dir.join("Cargo.toml").exists() {
+        let krate = read_crate(&dir.join("Cargo.toml"));
+        if CRATES_TO_PUBLISH
+            .iter()
+            .chain(CRATES_TO_AVOID_PUBLISH)
+            .any(|c| krate.name == *c)
+        {
+            dst.push(krate);
+        } else {
+            panic!("failed to find {:?} in whitelist or blacklist", krate.name);
+        }
+    }
+
+    for entry in dir.read_dir().unwrap() {
+        let entry = entry.unwrap();
+        if entry.file_type().unwrap().is_dir() {
+            find_crates(&entry.path(), dst);
+        }
+    }
+
+    fn read_crate(manifest: &Path) -> Crate {
+        let mut name = None;
+        let mut version = None;
+
+        // This is a "poor man's toml parser", but works well enough for our
+        // manifests.
+        for line in fs::read_to_string(manifest).unwrap().lines() {
+            if name.is_none() && line.starts_with("name = \"") {
+                name = Some(
+                    line.replace("name = \"", "")
+                        .replace("\"", "")
+                        .trim()
+                        .to_string(),
+                );
+            }
+            if version.is_none() && line.starts_with("version = \"") {
+                version = Some(
+                    line.replace("version = \"", "")
+                        .replace("\"", "")
+                        .trim()
+                        .to_string(),
+                );
+            }
+        }
+        let name = name.unwrap();
+        let version = version.unwrap();
+        let next_version = if CRATES_TO_PUBLISH.contains(&&name[..]) {
+            bump(&version)
+        } else {
+            version.clone()
+        };
+        Crate {
+            manifest: manifest.to_path_buf(),
+            name,
+            version,
+            next_version,
+        }
+    }
+}
+
+fn bump_version(krate: &Crate, crates: &[Crate]) {
+    let contents = fs::read_to_string(&krate.manifest).unwrap();
+
+    let mut new_manifest = String::new();
+    let mut is_deps = false;
+    for line in contents.lines() {
+        let mut rewritten = false;
+        if line.starts_with("version =") {
+            if CRATES_TO_PUBLISH.contains(&&krate.name[..]) {
+                println!(
+                    "bump `{}` {} => {}",
+                    krate.name, krate.version, krate.next_version
+                );
+                new_manifest.push_str(&line.replace(&krate.version, &krate.next_version));
+                rewritten = true;
+            }
+        }
+
+        is_deps = if line.starts_with("[") {
+            line.contains("dependencies")
+        } else {
+            is_deps
+        };
+
+        for other in crates {
+            if !is_deps || !line.starts_with(&format!("{} ", other.name)) {
+                continue;
+            }
+            if !line.contains(&other.version) {
```

Comment:
This is unlikely, but could you append and prepend `"` to the version string, so that `1.1.1` doesn't match `1.1.11`?